qoala/sim/scheduler.py

```diff
@@ -283,53 +342,183 @@ def wait(self, delta_time: float) -> Generator[EventExpression, None, None]:
+
+        new_cpu_tasks: Dict[int, TaskInfo] = {}
+        new_qpu_tasks: Dict[int, TaskInfo] = {}
+
+        # If there is a jump, jump to that block
+        if (
+            pid in self.host.interface.program_instance_jumps
+            and self.host.interface.program_instance_jumps[pid] != -1
+        ):
+            self._current_block_index[pid] = self.host.interface.program_instance_jumps[
+                pid
+            ]
+            self.host.interface.program_instance_jumps[pid] = -1
 
-        for batch in self._batches.values():
-            for inst in batch.instances:
-                all_tasks.append(inst.task_graph)
+        current_block_index = self._current_block_index[pid]
+        prog_instance = self.memmgr.get_process(pid).prog_instance
+        blocks = prog_instance.program.blocks
 
-        return all_tasks
+        # If it is not in range or cpu is not done with tasks with this pid
+        # We cannot add any more tasks because all block types have cpu tasks
+        if len(blocks) <= current_block_index or self.cpu_scheduler.task_exists_for_pid(
+            pid
+        ):
+            return
 
-    def get_tasks_to_schedule_for(self, batch_id: int) -> List[TaskGraph]:
-        all_tasks: List[TaskGraph] = []
+        block = prog_instance.program.blocks[current_block_index]
 
-        batch = self._batches[batch_id]
-        for inst in batch.instances:
-            all_tasks.append(inst.task_graph)
+        # If it is CL or CC it will only have tasks in CPU but others will have tasks in both
+        if block.typ in {
+            BasicBlockType.CL,
+            BasicBlockType.CC,
+        }:
+            graph = self._task_from_block_builder.build(
+                prog_instance, current_block_index, self._network_ehi
+            )
+            new_cpu_tasks.update(graph.get_tasks())
+
+            self._current_block_index[pid] += 1
+            # If scheduler does not have any task send a message to wake it up
+            self._comp.send_cpu_scheduler_message(Message(-1, -1, "New Task"))
+        elif not self.qpu_scheduler.task_exists_for_pid(pid):
+            # Note that we know that cpu does not have any tasks for this pid
+            graph = self._task_from_block_builder.build(
+                prog_instance, current_block_index, self._network_ehi
+            )
+            new_cpu_tasks.update(graph.partial_graph(ProcessorType.CPU).get_tasks())
+            new_qpu_tasks.update(graph.partial_graph(ProcessorType.QPU).get_tasks())
+
+            self._current_block_index[pid] += 1
+            # If scheduler does not have any task send a message to wake it up
+            self._comp.send_cpu_scheduler_message(Message(-1, -1, "New Task"))
+            self._comp.send_qpu_scheduler_message(Message(-1, -1, "New Task"))
+
+        self._cpu_scheduler.add_tasks(new_cpu_tasks)
+        self._qpu_scheduler.add_tasks(new_qpu_tasks)
+
+    def schedule_all(self) -> None:
```

Comment:
What exactly is the relationship between `schedule_next_for` and `schedule_all`? Could you add some more comments? There also seems to be quite some duplicate code. Is it possible to extract those parts and reuse them?